File: src/main/java/cope/nebula/client/feature/command/impl/FakePlayer.java
package cope.nebula.client.feature.command.impl;

import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import cope.nebula.client.events.ServerConnectionEvent;
import cope.nebula.client.events.ServerConnectionEvent.Type;
import cope.nebula.client.feature.command.Command;
import net.minecraft.client.entity.EntityOtherPlayerMP;
import net.minecraft.init.MobEffects;
import net.minecraft.potion.PotionEffect;
import net.minecraft.world.GameType;

public class FakePlayer extends Command {
    private static EntityOtherPlayerMP fakePlayer = null;

    public FakePlayer() {
        super("fakeplayer", LiteralArgumentBuilder.literal("fakeplayer")
                .executes((ctx) -> {
                    if (fakePlayer != null) {
                        // LogoutSpots testing:
                        // EVENT_BUS.post(new ServerConnectionEvent(Type.LEAVE, fakePlayer.getGameProfile(), fakePlayer));

                        mc.world.removeEntity(fakePlayer);
                        mc.world.removeEntityDangerously(fakePlayer);

                        fakePlayer = null;

                        send("Removed fake player.");
                        return 0;
                    }

                    fakePlayer = new EntityOtherPlayerMP(mc.world, mc.player.getGameProfile());
                    fakePlayer.copyLocationAndAnglesFrom(mc.player);
                    fakePlayer.inventory.copyInventory(mc.player.inventory);
                    fakePlayer.setHealth(20.0f);
                    fakePlayer.setGameType(GameType.SURVIVAL);

                    fakePlayer.addPotionEffect(new PotionEffect(MobEffects.RESISTANCE, 1, 999999));

                    mc.world.spawnEntity(fakePlayer);

                    send("Spawned fake player.");

                    return 0;
                }));
    }
}

File: src/main/kotlin/wtf/nebula/client/feature/command/impl/FakePlayer.kt
package wtf.nebula.client.feature.command.impl

import com.mojang.brigadier.Command.SINGLE_SUCCESS
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import net.minecraft.client.entity.EntityOtherPlayerMP
import net.minecraft.world.GameType
import wtf.nebula.client.feature.command.Command

class FakePlayer : Command(arrayOf("fakeplayer", "fp", "dummy", "dummyplayer"), "", "Spawns a fake player") {
    private var fakePlayer: EntityOtherPlayerMP? = null

    override fun build(builder: LiteralArgumentBuilder<Any>) {
        builder.executes {
            if (fakePlayer != null) {
                mc.world.removeEntityDangerously(fakePlayer)
                mc.world.removeEntity(fakePlayer)

                fakePlayer = null
            }

            else {
                val fake = EntityOtherPlayerMP(mc.world, mc.player.gameProfile)

                fake.copyLocationAndAnglesFrom(mc.player)
                fake.inventory.copyInventory(mc.player.inventory)
                fake.health = mc.player.health
                fake.absorptionAmount = mc.player.absorptionAmount
                fake.setGameType(GameType.SURVIVAL)

                mc.world.spawnEntity(fake)
                fakePlayer = fake
            }

            return@executes SINGLE_SUCCESS
        }
    }
}